promql/engine.go

```diff
@@ -559,6 +544,25 @@ func (ng *Engine) populateIterators(ctx context.Context, s *EvalStmt) (storage.Q
 	return querier, err
 }
 
+// extractFuncFromPath walks up the path and searches for the first instance of
+// a function or aggregation.
+func extractFuncFromPath(p []Node) string {
+	if len(p) == 0 {
+		return ""
+	}
+	switch n := p[len(p)-1].(type) {
+	case *AggregateExpr:
+		return n.Op.String()
+	case *Call:
+		return n.Func.Name
+	case *BinaryExpr:
+		// If we hit a binary expression we terminate since we only care about functions
+		// or aggregations over a single metric.
+		return ""
+	}
+	return extractFuncFromPath(p[:len(p)-1])
```

Comment:
This tail recursion can be converted to an iterated version.

Reply:
Go doesn't do tail call elimination, so this is less efficient and uses unnecessary stack memory.

Reply:
Since I read in chat that there was disagreement on this, adding one more opinion: I think this is fine either as a recursive or iterative variant. Performance or stack space shouldn't matter here. I also don't find the recursive variant confusing, so I think it's fine.